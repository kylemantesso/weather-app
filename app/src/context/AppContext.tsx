import React, { useEffect, useState } from "react";
import {
  fetchWeather,
  transformWeatherResult,
} from "../services/MetaweatherService";
import { IWeather, IWeatherResult } from "../types/weather";

export type FormatType = "C" | "F";

export interface IAppState {
  location: number;
  weather: IWeather;
  weatherResult: IWeatherResult;
  setLocation: (location: number) => void;
  formatType: FormatType;
  setFormatType: (formatType: FormatType) => void;
}

export const AppContext = React.createContext<Partial<IAppState>>({});

export const useAppContext = () => React.useContext(AppContext) as IAppState;

export const AppContextProvider: React.FC = ({ children }) => {
  const [location, setLocation] = useState<IAppState["location"]>();
  const [weather, setWeather] = useState<IAppState["weather"]>();
  const [weatherResult, setWeatherResult] = useState<
    IAppState["weatherResult"]
  >();
  const [formatType, setFormatType] = useState<IAppState["formatType"]>("C");

  // on location change, fetch weather data
  useEffect(() => {
    if (location) {
      fetchWeather(location)
        .then((weatherResult) => {
          setWeatherResult(weatherResult);
          setWeather(transformWeatherResult(weatherResult, formatType));
        })
        .catch((error) => {
          console.log(error);
          // TODO: Show nice error message and prompt for retry
        });
    }
  }, [location]);

  // on format type change, transform weather result
  useEffect(() => {
    if (weatherResult) {
      setWeather(transformWeatherResult(weatherResult, formatType));
    }
  }, [formatType]);

  return (
    <AppContext.Provider
      value={{
        location,
        setLocation,
        weather,
        formatType,
        setFormatType,
      }}
    >
      {children}
    </AppContext.Provider>
  );
};
