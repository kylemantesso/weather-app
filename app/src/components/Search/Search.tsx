import React, { useState } from "react";
import { Box, Select } from "grommet";
import { fetchLocations } from "../../services/MetaweatherService";
import { useAppContext } from "../../context/AppContext";
import { Spinner } from "../Spinner/Spinner";
import { ILocationResult } from "../../types/weather";

const defaultOptions: string[] = [];

export const Search = () => {
  const [options, setOptions] = useState(defaultOptions);
  const [locationQuery, setLocationQuery] = useState("");
  const [foundLocations, setFoundLocations] = useState<ILocationResult[]>([]);
  const [searching, setSearching] = useState(false);

  const { setLocation } = useAppContext();

  return (
    <Box
      align="center"
      justify="center"
      pad="large"
      direction="row"
      gap="medium"
    >
      <Select
        size="medium"
        placeholder="Serch location"
        value={locationQuery}
        options={options}
        onChange={({ option }) => {
          setSearching(false);
          setLocationQuery(option);
          setOptions([option]);
          if (foundLocations) {
            const locationId = foundLocations.find(
              (location) => location.title === option
            )!.woeid;
            if (locationId) {
              setLocation(locationId);
            }
          }
        }}
        onClose={() => setOptions([])}
        onSearch={(text) => {
          setSearching(true);
          if (text.length > 2) {
            fetchLocations(text)
              .then((locations) => {
                console.log(locations);
                const foundOptions = locations.map(
                  (location) => location.title
                );
                setFoundLocations(locations);
                setOptions(foundOptions);
                setSearching(false);
              })
              .catch((error) => {
                console.log(error);
                // TODO: Show nice error message and prompt for retry
              });
          }
        }}
      />
      <Box width="xxsmall">{searching && <Spinner />}</Box>
    </Box>
  );
};
