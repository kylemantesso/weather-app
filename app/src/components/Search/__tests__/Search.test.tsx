import React from "react";
import renderer from "react-test-renderer";

import * as AppContext from "../../../context/AppContext";
import { Search } from "../Search";

jest.mock("../../../context/AppContext");
const mockAppContext = AppContext as jest.Mocked<typeof AppContext>;
const mockSetLocation = jest.fn();
mockAppContext.useAppContext.mockReturnValue({
  setLocation: mockSetLocation,
});

describe("Search", () => {
  it("matches snapshot", () => {
    const component = renderer.create(<Search />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  // TODO: Add further tests for searching
});
