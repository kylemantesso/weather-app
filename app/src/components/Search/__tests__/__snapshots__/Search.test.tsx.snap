// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search matches snapshot 1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 dHbsaj"
>
  <button
    aria-label="Open Drop"
    className="StyledButton-sc-323bzc-0 bwJeiP Select__StyledSelectDropButton-sc-17idtfo-1 ksCGpf"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    type="button"
  >
    <div
      className="StyledBox-sc-13pk1d4-0 VvbUh"
    >
      <div
        className="StyledBox-sc-13pk1d4-0 bVYLzf"
      >
        <div
          className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 ksQrmr"
        >
          <input
            autoComplete="off"
            className="StyledTextInput-sc-1x30a0s-0 dWxGQe Select__SelectTextInput-sc-17idtfo-0 hWmVID"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Serch location"
            readOnly={true}
            size="medium"
            tabIndex="-1"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="StyledBox-sc-13pk1d4-0 iCDnTm"
        style={
          Object {
            "minWidth": "auto",
          }
        }
      >
        <svg
          aria-label="FormDown"
          className="StyledIcon-ofa7kd-0 kbXEEj"
          viewBox="0 0 24 24"
        >
          <polyline
            fill="none"
            points="18 9 12 15 6 9"
            stroke="#000"
            strokeWidth="2"
          />
        </svg>
      </div>
    </div>
  </button>
  <div
    className="StyledBox__StyledBoxGap-sc-13pk1d4-1 cxqgXz"
  />
  <div
    className="StyledBox-sc-13pk1d4-0 gybyyu"
  />
</div>
`;
